lib/ui/highlight-manager.js

```diff
@@ -9,36 +9,143 @@ function HighlightManager(opts) {
 
-    // TODO: maybe we can hook into the XBL mousemove handler and steal focus
-    // that way instead of manually setting a mousemove listener
-    this.popup.addEventListener('mousemove', this.adjustHighlight);
+    this.initMutationObserver();
+  },
+  initMutationObserver: function() {
+    // The existing code doesn't insert all the results at once; instead, rows
+    // are inserted after repeated timeouts, to keep the urlbar responsive (see
+    // _appendResultTimeout in autocomplete.xml). Each time a row is appended,
+    // the existing code reapplies the highlight to an item in the results
+    // list. So, as that happens, we want to continually steal the focus back.
+    // The simplest way to do this, it turns out, is with a MutationObserver.
+    // We listen for changes to the `url` attribute on the rows inside the
+    // richlistbox. This ensures changes are detected, even when rows are
+    // reused, rather than created (see _adjustAcItem in autocomplete.xml).
+    //
+    // TODO: Might want to throttle the callback if performance seems slow.
+    this.resultsObserver = new this.win.MutationObserver(this.mutationHandler);
+    const results = this.win.document.getAnonymousElementByAttribute(this.popup, 'anonid', 'richlistbox');
+    const resultsObserverConfig = {
+      childList: true,
+      subtree: true,
+      attributes: true,
+      attributeFilter: ['url']
+    };
+    this.resultsObserver.observe(results, resultsObserverConfig);
   },
   destroy: function() {
-    this.popup.removeEventListener('mousemove', this.adjustHighlight);
-    this.events.unsubscribe('recommendation-shown', this.adjustHighlight);
+    this.resultsObserver.disconnect();
+    this.events.unsubscribe('recommendation-shown', this.stealHighlight);
     this.events.unsubscribe('navigational-key', this.adjustHighlight);
     delete this.popup;
     delete this.recommendationView;
     delete this.win;
   },
+  mutationHandler: function() {
+    if (this.recommendation.el && !this.recommendation.el.collapsed) {
+      this.stealHighlight();
+    }
+  },
+  stealHighlight: function() {
+    // When the recommendation has just been shown, if the default list item is
+    // highlighted, steal the highlight and set it on the recommendation.
+    //
+    // The existing XBL code wraps each insertion in a setTimeout, so we can't
+    // rely on setTimeout to steal the highlight: if 30 row rendering calls are
+    // in the event loop ahead of this code, the blue highlight will very
+    // noticeably flicker on the top result row. To avoid this, we steal the
+    // highlight immediately, then use rAF to steal it just before the next
+    // paint, then set *another* rAF inside the first rAF, preventing another
+    // paint from showing a result in the list as highlighted. This seems a
+    // little excessive, but actually the inner rAF seems to be the key to
+    // truly minimizing the flickering as we steal the highlight.
```

Comment:
It is the height of madness, that's for sure. I guess the time when the setTimeouts will be executed is unpredictable, so if we spray the event loop with enough focus stealing DOM changes, we wind up doing pretty well on average?